common/txmgr/abandoned_tracker.go

```diff
@@ -0,0 +1,194 @@
+	BLOCK_HASH types.Hashable,
+	R txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH],
+	SEQ types.Sequence,
+	FEE feetypes.Fee,
+] struct {
+	txStore *txmgrtypes.TxStore[ADDR, CHAIN_ID, TX_HASH, BLOCK_HASH, R, SEQ, FEE]
+	ks      *txmgrtypes.KeyStore[ADDR, CHAIN_ID, SEQ]
+	client  *txmgrtypes.TxmClient[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]
+	chainID CHAIN_ID
+	lggr    logger.Logger
+	txes    []AbandonedTx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]
+}
+
+// NewAbandonedTracker creates a new AbandonedTracker
+func NewAbandonedTracker[
+	CHAIN_ID types.ID,
+	ADDR types.Hashable,
+	TX_HASH types.Hashable,
+	BLOCK_HASH types.Hashable,
+	R txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH],
+	SEQ types.Sequence,
+	FEE feetypes.Fee,
+](
+	txStore *txmgrtypes.TxStore[ADDR, CHAIN_ID, TX_HASH, BLOCK_HASH, R, SEQ, FEE],
+	keystore *txmgrtypes.KeyStore[ADDR, CHAIN_ID, SEQ],
+	client *txmgrtypes.TxmClient[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE],
+	chainID CHAIN_ID,
+	lggr logger.Logger,
+) AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE] {
+	return AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]{
+		txStore: txStore,
+		ks:      keystore,
+		chainID: chainID,
+		client:  client,
+		lggr:    lggr.Named("AbandonedTracker"),
+		txes:    make([]AbandonedTx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], 0),
+	}
+}
+
+func (tracker *AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) containsTx(
+	tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) bool {
+	for _, atx := range tracker.txes {
+		if atx.tx.ID == tx.ID {
+			return true
+		}
+	}
+	return false
+}
+
+// insertAbandonedTx inserts a transaction into the tracker
+func (tracker *AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) insertAbandonedTx(
+	tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) {
+	if tracker.containsTx(tx) {
+		return
+	}
+
+	tracker.lggr.Debugw(fmt.Sprintf("inserting tx %v", tx.ID))
+	tracker.txes = append(tracker.txes, AbandonedTx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{
+		tx:        tx,
+		fatalTime: time.Now().Add(TTL),
+	})
+}
+
+// markFatal sets a transaction's state to fatal_error
+func (tracker *AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) markFatal(
+	ctx context.Context,
+	atx AbandonedTx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) {
+	tracker.lggr.Infow(fmt.Sprintf("tx %v marked as fatal for exceeding ttl", atx.tx.ID))
+
+	atx.tx.Error.SetValid(AbandonedErrorMsg)
+
+	err := (*tracker.txStore).UpdateTxFatalError(ctx, atx.tx)
+	if err != nil {
+		tracker.lggr.Errorw(fmt.Sprintf("failed to mark tx %v as fatal", atx.tx.ID))
+		// TODO: Handle error
+	}
+}
+
+// getAbandonedAddresses retrieves fromAddress’s in evm.txes that are not present in the Confirmer's enabledAddresses list
+func (tracker *AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) getAbandonedAddresses(enabledAddrs []ADDR) ([]ADDR, error) {
```

Comment:
This functions could be called only once, when the enabledAddresses list is changed, by the Confirmer.
And then, on every new head, you just track these abandoned txs.

Reply:
Is it not possible to have more transactions become abandoned over time? Or is this only necessary on startup to check the database for all abandoned transactions, and then we don't have to worry about more occurring?